docs/specs/mlnet-cli/MLNET-CLI-Specs.md

```diff
@@ -0,0 +1,969 @@
+
+# Who is the Customer
+
+- Regular .NET developers getting started with machine learning while trying to use .NET (C# and F# most of all) for ML.
+
+- Specific developer roles are: enterprise developers, start-up developers, ISV developers and internal MSFT teams developers.
+
+
+# Goals
+
+The business goals are the following, depending on the possible scenarios:
+
+- **"New to ML" scenario: (Previews)** Regular .NET developer with minimum or no data science skills is able to go from raw data to working model.
+
+- **"Productivity plus quality" scenario: (Previews)** .NET developers with good data science skills is able to significantly improve productivity when creating a model while making sure they get the best model even when they might know the label, features to use, etc. but need to maximize accuracy, infer speed and other metrics without spending much time on custom low-level experiments.
+
+- **"ML.NET learning and customization" scenario: (Previews)** .NET developers are able to research the generated pipeline's implementation code (C# + ML.NET API) for customization of the generated ML.NET code.
+
+- **"Cross-platform" scenario: (Previews)** .NET developers are able to use the ML.NET CLI from the same OS supported by .NET Core: Windows, Mac and Linux, so they are not tight to a single environment like Windows and VisualStudio but can be used from any OS CLI and code editor.
+
+- **"Advanced ML learning and customization" scenario: (v0.2)** Developers with ML skills are able to customize internal ML parameters such as trainerexclusion/inclusion lists, specific featurization, metrics to prioritize, etc. for more advanced learning and customization of the generated models and ML.NET code and even to improve total execution times by providing certain prescriptive ML "shortcuts".
+
+- **Azure integration for scaling-out (v1.0 or vNext)**: In next versions, an important goal will be to be able to scale-out the explorations and training in parallel tasks deployed into Azure in addition to running it on a local PC. AutoML in Azure is a great example, but regular training of a model could also be performed in Azure.
+
+The goals for the features is to automate the following steps when building a model to achieve the above business goals:
+
+**Foundational features:** 
+
+- Provide an end-to-end **ML.NET CLI** for developers (i.e. *"mlnet new"*) to generate either the final trained model and the pipeline's C#/ML.NET implementation code in a similar fashion to the [.NET Core CLI](https://docs.microsoft.com/en-us/dotnet/core/tools/?tabs=netcore2x). The CLI is also a foundation upon which higher-level tools, such as Integrated Development Environments (IDEs) can rest.
```

Comment:
>mlnet new [](start = 62, length = 9)

Function should do one thing and do it well, thus
mlnet new -> generates code
mlnet train -> trains a model.

Reply:
This is something to be discussed and validated/invalidated by users. My poit of view is that a "mlnet new" should generate the scoring application code with anything it needs to run (end-to-end) which means to have the .ZIP file, already. And since that was already created by AutoML, it can be provided without the user needing to do an additional step (train the model).
I think we need to think about what .NET developers new to ML.NET would want for their usual workflow more than structure everything in very granular operations "Function should do one thing and do it well". The [SRP (Single Responsability Principle)](https://en.wikipedia.org/wiki/Single_responsibility_principle) applies to classes and methods, not necessarily to a CLI which should accommodate to the user's workflow most of all.